src/core/thread/src_match_controller.cpp

```diff
@@ -54,12 +54,21 @@ SourceMatchController::SourceMatchController(Instance &aInstance)
 
 void SourceMatchController::IncrementMessageCount(Child &aChild)
 {
+    if (aChild.GetIndirectMessageCount() == aChild.GetIndirectMaxMessageCount())
+    {
+        otLogWarnMac("IncrementMessageCount(child 0x%04x) called when already at maximum count.", aChild.GetRloc16());
+        ExitNow();
```

Comment:
Hi. Thanks for contributing to OpenThread.

Regarding this check, I am not sure if we want to let the message counter saturate at certain max value. For correct behavior, we need to keep track of exact number of messages, otherwise the src macth controller logic won't function correctly.

My suggestion is instead of this change we use `OT_STATIC_ASSERT` to ensure we can never get to such a situation.
- The maximum number of messages we can possibly have in the queue is limited by total number of message buffers or `OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS` config option (which by default is 44)
- The `mQueuedMessageCount`  in `Child` is 13-bit which means we can go up 8192 (which seems to be good high limit for this

To make sure we don't go over limit, my suggestion is that to add a static-assert like this in `topology.hpp`
```
OT_STATIC_ASSERT(OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS < 8192, "mQueuedMessageCount cannot fit max required");` 
```

Reply:
Fixed to static assertion.